### PR TITLE
Create a signal to unassign roles from user when the role is deleted

### LINKED ISSUE #369: User's role field needs to have a cascade delete
#357 replaced the FK with an `ArrayField`. This introduced a problem of the roles table falling out of sync with the users table. I encountered an issue when trying to sync a user with the bot. One of their roles had been deleted from the roles table, but it was still present in their roles field. The validator ensures all roles exist in the table, so the request was failing.

Apparently [Django Postgres Extensions](https://django-postgres-extensions.readthedocs.io/en/latest/arrays.html) have a solution for this.

### PR DESCRIPTION
**Pull request by D0rs4n**
Add a signal to the api app that automatically unassigns
deleted roles from users that have them
Closes #369

### CODE REVIEW

File: pydis_site/apps/api/signals.py
Comment (Bluenix2) [inline pydis_site/apps/api/signals.py:12]: Is there not a way to do this in batch? We should let SQL do this

Comment (D0rs4n) [reply pydis_site/apps/api/signals.py:12]: Oh, I see your concern. That wasn't a wise choice indeed. Although, I think there's a way to solve this using Django ORM.

Comment (D0rs4n) [reply pydis_site/apps/api/signals.py:12]: It seems like, the Django ORM, F objects in particular does not support ArrayFields. Hence that's the only way to solve it via the Django ORM, for now.
However, I've never really worked with arrayfields using raw SQL.

Comment (jchristgit) [reply pydis_site/apps/api/signals.py:12]: This isn't possible with plain PostgreSQL, except if we were to use triggers, and I would currently stay away from that personally.

The original approach here was the correct way: foreign key constraints. But since this got removed due to performance issues (waiting on reply from a Django expert friend on this topic), this is a fine approach.

**Review by jchristgit — APPROVED** [review]
This looks good, thank you!

Could you rebase off master, please?

**Comment by jchristgit** [comment]
Hi @Akarys42, could you give this a second review?

**Review by ChrisLovering — APPROVED** [review]
The seems to work well locally 👌

**Pull request merged**
